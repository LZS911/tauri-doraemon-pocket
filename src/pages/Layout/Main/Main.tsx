import { Breadcrumb } from 'antd';
import React, { ReactNode, useMemo } from 'react';
import { useLocation } from 'react-router-dom';

const Main: React.FC<{
  children: ReactNode;
}> = ({ children }) => {
  const { pathname } = useLocation();
  const breadcrumbList = useMemo(() => {
    return pathname.slice(1).split('/');
  }, [pathname]);

  return (
    <main
      className={`min-h-[100vh] overflow-auto bg-slate-100 p-6 dark:!bg-black`}
    >
      {breadcrumbList.length > 1 && (
        <Breadcrumb
          className="mb-2"
          separator={<span className="opacity-50 dark:text-white">/</span>}
        >
          {breadcrumbList.map((v, index) => {
            return index === breadcrumbList.length - 1 ? (
              <Breadcrumb.Item key={v}>
                <span className="dark:text-white">{v}</span>
              </Breadcrumb.Item>
            ) : (
              <Breadcrumb.Item key={v}>
                <span className="opacity-50 dark:!text-white">{v}</span>
              </Breadcrumb.Item>
            );
          })}
        </Breadcrumb>
      )}

      {children}
    </main>
  );
};

export default Main;
