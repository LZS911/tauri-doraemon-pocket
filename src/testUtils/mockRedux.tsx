import { configureStore } from '@reduxjs/toolkit';
import { combineReducers } from 'redux';
import layout from '../store/layout';
import config from '../store/config';
import { Dictionary } from '../typing/common.type';
import { Provider } from 'react-redux';
import React, { ReactNode } from 'react';
const reducers = combineReducers({
  layout,
  config,
});
export const storeFactory: any = (initStore: Dictionary = {}) => {
  return configureStore({
    reducer: reducers,
    preloadedState: initStore,
  });
};

export const CustomProvider: React.FC<{
  initStore: Dictionary;
  children: ReactNode;
}> = (props) => {
  return (
    <Provider store={storeFactory(props.initStore)}>{props.children}</Provider>
  );
};
