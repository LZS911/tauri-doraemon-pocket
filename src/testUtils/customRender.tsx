/* eslint-disable @typescript-eslint/no-explicit-any */
import { render, renderHook } from '@testing-library/react';
import { Provider } from 'react-redux';
import {
  BrowserRouter,
  MemoryRouter,
  MemoryRouterProps,
  Router,
  RouterProps,
} from 'react-router-dom';
import { Dictionary } from '../typing/common.type';
import { storeFactory } from './mockRedux';
type RenderParams = Parameters<typeof render>;

export const renderWithRedux: (arg: any) => ReturnType<typeof render> = (
  ...[ui, option, initStore]: [...RenderParams, Dictionary?]
) => {
  return render(
    <Provider store={storeFactory(initStore)}>{ui}</Provider>,
    option
  );
};
export const renderWithRouter: any = (...[ui, option]: [...RenderParams]) => {
  return render(<BrowserRouter>{ui}</BrowserRouter>, option);
};

export const renderHooksWithRedux = <TProps, TResult>(
  hooks: (props: TProps) => TResult,
  storeState: Dictionary = {}
) => {
  return renderHook(hooks, {
    wrapper: Provider,
    initialProps: {
      store: storeFactory(storeState),
    },
  } as any);
};

export const renderWithServerRouter: any = (
  ...[ui, option, props]: [...RenderParams, RouterProps]
) => {
  return render(<Router {...props}>{ui}</Router>, option);
};

export const renderWithServerRouterAndRedux: any = (
  ...[ui, option, props]: [...RenderParams, RouterProps]
) => {
  return render(
    <Router {...props}>
      {' '}
      <Provider store={storeFactory()}>{ui}</Provider>,
    </Router>,
    option
  );
};

export const renderWithMemoryRouter: any = (
  ...[ui, option, props]: [...RenderParams, MemoryRouterProps?]
) => {
  return render(<MemoryRouter {...props}>{ui}</MemoryRouter>, option);
};
