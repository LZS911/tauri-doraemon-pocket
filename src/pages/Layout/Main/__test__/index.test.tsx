import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { renderWithRouter } from '../../../../testUtils/customRender';
import Main from '../index';

const children = <>children</>;

describe('test Main', () => {
  test('should match snapshot', () => {
    const { container } = renderWithRouter(<Main>{children}</Main>);
    expect(container).toMatchSnapshot();
  });

  test('should show breadcrumb by router', () => {
    render(
      <MemoryRouter initialEntries={['/component/table']}>
        <Main>{children}</Main>
      </MemoryRouter>
    );

    // expect(screen.queryByText('component')).toBeInTheDocument();
    // expect(screen.queryByText('table')).toBeInTheDocument();
  });
});
