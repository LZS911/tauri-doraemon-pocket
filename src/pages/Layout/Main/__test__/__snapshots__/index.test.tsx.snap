// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Main should match snapshot 1`] = `
<div>
  <main
    class="min-h-[calc(100vh-50px)] overflow-auto bg-slate-100 p-6 dark:!bg-black"
  >
    children
  </main>
</div>
`;
